### PR TITLE
Wrap ZoneProperty:UserViewFactors:bySurfaceName object

### LINKED ISSUE #3664: Wrap ZoneProperty:UserViewFactors:bySurfaceName object
https://bigladdersoftware.com/epx/docs/9-1/input-output-reference/group-advanced-surface-concepts.html#zonepropertyuserviewfactorsbysurfacename

@yzhou601 Can you attach one of your IDF files that uses this object?

### PR DESCRIPTION
**Pull request by jmarrec**
Supersedes #3671 (essential work by @joseph-robertson). Most of the discussion/review is on #3671.

Fix #3664.

Design doc: https://docs.google.com/document/d/1P8O6GRQAZda5kK_2Qu2CJRkBGUxTIhJwvJ0-JNpTU1A/edit#heading=h.c285d1mdhew5

TODO:
- [x] Update OpenStudio.idd
- [x] Create source code files
- [x] Update all cmakelists
- [x] Create FT
- [x] Create RT
- [x] Check ostream on view factor
- [x] Create test for RT
- [x] Check RT in app
- [x] Create unit tests
- [x] Create regression tests
- [x] Check FT
- [x] Check simulation runs successfully

### CODE REVIEW

File: openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+      OptionalDouble _viewFactor = workspaceGroup.getDouble(ZoneProperty_UserViewFactors_bySurfaceNameExtensibleFields::ViewFactor);
+
+      // If the objects translated ok, and the viewFactor double value exists
+      if (fromModelObject && toModelObject && _viewFactor) {
+        // try to create a viewFactor, which will throw in case wrong type, or wrong double value. Here we let it slide though
+        try {
+          ViewFactor viewFactor(fromModelObject.get(), toModelObject.get(), _viewFactor.get());
+          // add the view factor
+          if (zoneProp.addViewFactor(viewFactor)) {
```
Comment (joseph-robertson) [inline openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: Did you try out the RT in OSApp? Should this be `!zoneProp.addViewFactor(viewFactor)`? Seems like this warning is thrown if the viewFactor is added.

File: openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            LOG(Warn, "Adding ViewFactor in ThermalZone " << thermalZone.nameString()
+                  << " failed for viewFactor=" << viewFactor << ".");
+          }
+        } catch (...) {
+          // The ViewFactor Ctor threw, so there's a wrong type, or a wrong double
+          LOG(Error, "Couldn't not create ViewFactor in ThermalZone " << thermalZone.nameString()
```
Comment (joseph-robertson) [inline openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: Remove "not"?

File: openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp [hunk 3]
```diff
@@ -648,19 +651,203 @@ TEST_F(EnergyPlusFixture, ReverseTranslator_ScheduleFile) {
...
+
+  openstudio::IdfObject idf_zoneProp(openstudio::IddObjectType::ZoneProperty_UserViewFactors_bySurfaceName);
+  idf_zoneProp.setString(0, "Thermal Zone 1"); // Zone or ZoneList Name
+  idf_zoneProp.setString(1, "Surface 1"); // From Surface 1
+  idf_zoneProp.setString(2, "Surface 2"); // To Surface 2
+  idf_zoneProp.setDouble(3, 0.25); // View Factor 1
+
+  openstudio::WorkspaceObject epZonePropertyUserViewFactorsBySurfaceName = workspace.addObject(idf_zoneProp).get();
+
+  workspace.save("/media/Data/Software/Others/OS-build-develop/test_rt.idf", true);
```
Comment (joseph-robertson) [inline openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp]: Is this a debug thing?

File: openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp [hunk 4]
```diff
@@ -648,19 +651,203 @@ TEST_F(EnergyPlusFixture, ReverseTranslator_ScheduleFile) {
...
+  EXPECT_EQ(zoneProp.thermalZone().name().get(), "Thermal Zone 1 Thermal Zone");
+
+  // TODO: JM 2019-09-13 NOT SURE WHETHER WE WANT TO ALLOW TO==FROM here
+  bool isToEqualFromAllowed = false;
+  if (!isToEqualFromAllowed) {
+    // If we don't allow toSurface to equal to fromSurface
+    EXPECT_EQ(0u, zoneProp.numberofViewFactors());
+
+  } else {
+    // If we allow toSurface to be equal to fromSurface
```
Comment (joseph-robertson) [inline openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp]: We just want this block right?

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+ModelObject ViewFactor::toSurface() const {
+  return m_to_surface;
+}
+
+double ViewFactor::viewFactor() const {
+  return m_view_factor;
+}
+
+std::ostream& operator<< (std::ostream& out, const openstudio::model::ViewFactor& viewFactor) {
+  out << "(from " << viewFactor.fromSurface().iddObject().name() << "='" << viewFactor.fromSurface().nameString() << "', "
```
Comment (joseph-robertson) [inline openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: Yes, this is much cleaner.

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    boost::optional<ThermalZone> thermalZone = getObject<ModelObject>().getModelObjectTarget<ThermalZone>(OS_ZoneProperty_UserViewFactors_BySurfaceNameFields::ThermalZoneName);
+    OS_ASSERT(thermalZone);
+    return thermalZone.get();
+  }
+
+  unsigned int ZonePropertyUserViewFactorsBySurfaceName_Impl::numberofViewFactors() const {
+    return numExtensibleGroups();
+  }
+
+  boost::optional<unsigned> ZonePropertyUserViewFactorsBySurfaceName_Impl::viewFactorIndex(const ViewFactor& t_viewFactor) const {
```
Comment (joseph-robertson) [inline openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: Nice addition.

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+
+    // If found, we compute the index by using std::distance between the start of vector and the iterator returned by std::find_if
+    if (it != currentViewFactors.end()) {
+      result = std::distance(currentViewFactors.begin(), it);
+    }
+
+    return result;
+  }
+
+  boost::optional<ViewFactor> ZonePropertyUserViewFactorsBySurfaceName_Impl::getViewFactor(unsigned groupIndex) const {
```
Comment (joseph-robertson) [inline openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: Nice addition.

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+          << "'is not part of the ThermalZone.");
+      return false;
+    }
+
+    // Check if viewFactor already exists
+    boost::optional<unsigned> _existingIndex = viewFactorIndex(viewFactor);
+    if (_existingIndex) {
+      boost::optional<ViewFactor> _viewFactor = getViewFactor(_existingIndex.get());
+      OS_ASSERT(_viewFactor);
+      LOG(Warn, "For " << briefDescription() << ", ViewFactor already exists, will be modified in place from " << _viewFactor.get()
```
Comment (joseph-robertson) [inline openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: Do we want the view factor to be overwritten, or ignored (and the existing one is left untouched)?

@shorowit

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+        result.push_back(_viewFactor.get());
+      } else {
+        LOG(Error, briefDescription() << " has an invalid ViewFactor group at index " << i);
+      }
+    }
+
+    return result;
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactors(const std::vector<ViewFactor> &viewFactors) {
```
Comment (joseph-robertson) [inline openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: So this method would return false if at least one view factor is not added (but the rest are added)? Or maybe the "Continuing with others." is throwing me- it's not actually continuing to add the remaining view factors, right?

File: openstudiocore/src/model/test/ZonePropertyUserViewFactorsBySurfaceName_GTest.cpp [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+  EXPECT_THROW(ViewFactor(s, intMass, 1.2), openstudio::Exception);
+
+  // Wrong type
+  EXPECT_THROW(ViewFactor(s, z, 0.9), openstudio::Exception);
+  EXPECT_THROW(ViewFactor(z, s, 0.9), openstudio::Exception);
+
+
+  // TODO: JM 2019-09-13 NOT SURE WHETHER WE WANT TO ALLOW TO==FROM here
+  // Test that you cannot add a view factor  if toSurface == fromSurface
+  EXPECT_THROW(ViewFactor(s, s, 0.25),  openstudio::Exception);
```
Comment (joseph-robertson) [inline openstudiocore/src/model/test/ZonePropertyUserViewFactorsBySurfaceName_GTest.cpp]: Will want to change this to no throw.

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp (hunk 8 above)
Comment (shorowit) [reply openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: I'd overwrite it if it's not too much work.

Comment (joseph-robertson) [reply openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: OK. That's how it's written currently.

File: openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp (hunk 2 above)
Comment (jmarrec) [reply openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: No Keep it here

File: openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp (hunk 1 above)
Comment (jmarrec) [reply openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: Yes please change that. Good catch

File: openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp (hunk 4 above)
Comment (jmarrec) [reply openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp]: Yes i thought i cleaned that up..

File: openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp (hunk 2 above)
Comment (shorowit) [reply openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: Should either be "Couldn't create" or "Could not create"

File: openstudiocore/src/model/test/ZonePropertyUserViewFactorsBySurfaceName_GTest.cpp (hunk 10 above)
Comment (jmarrec) [reply openstudiocore/src/model/test/ZonePropertyUserViewFactorsBySurfaceName_GTest.cpp]: It looks like i forgot to push my last commit... damn! This is on work machine that's off and i dont go back until Monday

File: openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp (hunk 3 above)
Comment (jmarrec) [reply openstudiocore/src/energyplus/Test/ReverseTranslator_GTest.cpp]: Definitiely...

File: openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp (hunk 2 above)
Comment (joseph-robertson) [reply openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: Yes, that's what I meant.

Comment (jmarrec) [reply openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp]: Yes, read it too quickly (just like I wrote it too quickly...). This should not be a double negation, just express that it was not possible. Another brick in the wall:

> We don't need no education...

**Comment by tijcolem** [comment]
This may already known and being addressed but CI reported a couple of new unit test failures. I replayed it to make sure these are not intermittent failures, but these failed on the second attempt (see below).

EnergyPlusFixture.ReverseTranslator_ZonePropertyUserViewFactorsBySurfaceName_SameSurface (Failed)
ModelFixture.ZonePropertyUserViewFactorsBySurfaceName_ViewFactor (Failed)

File: openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp (hunk 9 above)
Comment (jmarrec) [reply openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp]: This is a choice really, but I thought if someone goes wrong you might want to return `false` in addition to logging a warn (or error) (as in "make sure you got what you want out of this method")

**Comment by jmarrec** [comment]
@tijcolem Yeah it's my fault, I forgot to push my latest commit to the remote on Friday. Joe cleaned it up but there was one more thing that needed change (cf https://github.com/NREL/OpenStudio/pull/3675/commits/a544b86fb97b388f40a6a7b1896e13047e6add5c#diff-4523b9e8d5a4310f0b4e569c01353082R71). I merged my commit with Joe's, should be good to go. I'll check CI result to be sure, but locally everything is fine.

I'll report back, so you can merge if OK (I can't merge stuff to develop)

**Comment by jmarrec** [comment]
Ok all good @tijcolem you can hit merge

**Pull request merged**